// standardized-audio-context/build/esm/factories/analyser-node-constructor.js
import * as tslib_1 from "tslib";
import { getNativeContext } from '../helpers/get-native-context';
var DEFAULT_OPTIONS = {
    channelCount: 2,
    channelCountMode: 'max',
    channelInterpretation: 'speakers',
    fftSize: 2048,
    maxDecibels: -30,
    minDecibels: -100,
    smoothingTimeConstant: 0.8
};
export var createAnalyserNodeConstructor = function (createAnalyserNodeRenderer, createNativeAnalyserNode, isNativeOfflineAudioContext, noneAudioDestinationNodeConstructor) {
    return /** @class */ (function (_super) {
        tslib_1.__extends(AnalyserNode, _super);
        function AnalyserNode(context, options) {
            if (options === void 0) { options = DEFAULT_OPTIONS; }
            var _this = this;
            var nativeContext = getNativeContext(context);
            var mergedOptions = tslib_1.__assign({}, DEFAULT_OPTIONS, options);
            var nativeNode = createNativeAnalyserNode(nativeContext, mergedOptions);
            var analyserNodeRenderer = (isNativeOfflineAudioContext(nativeContext)) ? createAnalyserNodeRenderer() : null;
            _this = _super.call(this, context, nativeNode, analyserNodeRenderer) || this;
            _this._nativeNode = nativeNode;
            return _this;
        }
        Object.defineProperty(AnalyserNode.prototype, "fftSize", {
            get: function () {
                return this._nativeNode.fftSize;
            },
            set: function (value) {
                this._nativeNode.fftSize = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(AnalyserNode.prototype, "frequencyBinCount", {
            get: function () {
                return this._nativeNode.frequencyBinCount;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(AnalyserNode.prototype, "maxDecibels", {
            get: function () {
                return this._nativeNode.maxDecibels;
            },
            set: function (value) {
                this._nativeNode.maxDecibels = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(AnalyserNode.prototype, "minDecibels", {
            get: function () {
                return this._nativeNode.minDecibels;
            },
            set: function (value) {
                this._nativeNode.minDecibels = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(AnalyserNode.prototype, "smoothingTimeConstant", {
            get: function () {
                return this._nativeNode.smoothingTimeConstant;
            },
            set: function (value) {
                this._nativeNode.smoothingTimeConstant = value;
            },
            enumerable: true,
            configurable: true
        });
        AnalyserNode.prototype.getByteFrequencyData = function (array) {
            this._nativeNode.getByteFrequencyData(array);
        };
        AnalyserNode.prototype.getByteTimeDomainData = function (array) {
            this._nativeNode.getByteTimeDomainData(array);
        };
        AnalyserNode.prototype.getFloatFrequencyData = function (array) {
            this._nativeNode.getFloatFrequencyData(array);
        };
        AnalyserNode.prototype.getFloatTimeDomainData = function (array) {
            this._nativeNode.getFloatTimeDomainData(array);
        };
        return AnalyserNode;
    }(noneAudioDestinationNodeConstructor));
};
