// standardized-audio-context/build/esm/factories/audio-buffer-source-node-renderer-factory.js
var _this = this;
import * as tslib_1 from "tslib";
import { getNativeNode } from '../helpers/get-native-node';
import { isOwnedByContext } from '../helpers/is-owned-by-context';
import { renderInputsOfAudioNode } from '../helpers/render-inputs-of-audio-node';
export var createAudioBufferSourceNodeRendererFactory = function (createNativeAudioBufferSourceNode) {
    return function () {
        var nativeNode = null;
        var start = null;
        var stop = null;
        return {
            set start(value) {
                start = value;
            },
            set stop(value) {
                stop = value;
            },
            render: function (proxy, offlineAudioContext) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                var options;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (nativeNode !== null) {
                                return [2 /*return*/, nativeNode];
                            }
                            nativeNode = getNativeNode(proxy);
                            // If the initially used nativeNode was not constructed on the same OfflineAudioContext it needs to be created again.
                            if (!isOwnedByContext(nativeNode, offlineAudioContext)) {
                                options = {
                                    buffer: nativeNode.buffer,
                                    channelCount: nativeNode.channelCount,
                                    channelCountMode: nativeNode.channelCountMode,
                                    channelInterpretation: nativeNode.channelInterpretation,
                                    detune: 0,
                                    loop: nativeNode.loop,
                                    loopEnd: nativeNode.loopEnd,
                                    loopStart: nativeNode.loopStart,
                                    playbackRate: nativeNode.playbackRate.value
                                };
                                nativeNode = createNativeAudioBufferSourceNode(offlineAudioContext, options);
                                if (start !== null) {
                                    nativeNode.start.apply(nativeNode, start);
                                }
                                if (stop !== null) {
                                    nativeNode.stop(stop);
                                }
                            }
                            return [4 /*yield*/, renderInputsOfAudioNode(proxy, offlineAudioContext, nativeNode)];
                        case 1:
                            _a.sent();
                            return [2 /*return*/, nativeNode];
                    }
                });
            }); }
        };
    };
};
