// standardized-audio-context/build/esm/factories/biquad-filter-node-renderer-factory.js
var _this = this;
import * as tslib_1 from "tslib";
import { connectAudioParam } from '../helpers/connect-audio-param';
import { getNativeNode } from '../helpers/get-native-node';
import { isOwnedByContext } from '../helpers/is-owned-by-context';
import { renderAutomation } from '../helpers/render-automation';
import { renderInputsOfAudioNode } from '../helpers/render-inputs-of-audio-node';
export var createBiquadFilterNodeRendererFactory = function (createNativeBiquadFilterNode) {
    return function () {
        var nativeNode = null;
        return {
            render: function (proxy, offlineAudioContext) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                var options;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (nativeNode !== null) {
                                return [2 /*return*/, nativeNode];
                            }
                            nativeNode = getNativeNode(proxy);
                            if (!!isOwnedByContext(nativeNode, offlineAudioContext)) return [3 /*break*/, 5];
                            options = {
                                Q: nativeNode.Q.value,
                                channelCount: nativeNode.channelCount,
                                channelCountMode: nativeNode.channelCountMode,
                                channelInterpretation: nativeNode.channelInterpretation,
                                detune: nativeNode.detune.value,
                                frequency: nativeNode.frequency.value,
                                gain: nativeNode.gain.value,
                                type: nativeNode.type
                            };
                            nativeNode = createNativeBiquadFilterNode(offlineAudioContext, options);
                            return [4 /*yield*/, renderAutomation(proxy.context, offlineAudioContext, proxy.Q, nativeNode.Q)];
                        case 1:
                            _a.sent();
                            return [4 /*yield*/, renderAutomation(proxy.context, offlineAudioContext, proxy.detune, nativeNode.detune)];
                        case 2:
                            _a.sent();
                            return [4 /*yield*/, renderAutomation(proxy.context, offlineAudioContext, proxy.frequency, nativeNode.frequency)];
                        case 3:
                            _a.sent();
                            return [4 /*yield*/, renderAutomation(proxy.context, offlineAudioContext, proxy.gain, nativeNode.gain)];
                        case 4:
                            _a.sent();
                            return [3 /*break*/, 10];
                        case 5: return [4 /*yield*/, connectAudioParam(proxy.context, offlineAudioContext, proxy.Q)];
                        case 6:
                            _a.sent();
                            return [4 /*yield*/, connectAudioParam(proxy.context, offlineAudioContext, proxy.detune)];
                        case 7:
                            _a.sent();
                            return [4 /*yield*/, connectAudioParam(proxy.context, offlineAudioContext, proxy.frequency)];
                        case 8:
                            _a.sent();
                            return [4 /*yield*/, connectAudioParam(proxy.context, offlineAudioContext, proxy.gain)];
                        case 9:
                            _a.sent();
                            _a.label = 10;
                        case 10: return [4 /*yield*/, renderInputsOfAudioNode(proxy, offlineAudioContext, nativeNode)];
                        case 11:
                            _a.sent();
                            return [2 /*return*/, nativeNode];
                    }
                });
            }); }
        };
    };
};
