// standardized-audio-context/build/esm/factories/oscillator-node-renderer-factory.js
var _this = this;
import * as tslib_1 from "tslib";
import { connectAudioParam } from '../helpers/connect-audio-param';
import { getNativeNode } from '../helpers/get-native-node';
import { isOwnedByContext } from '../helpers/is-owned-by-context';
import { renderAutomation } from '../helpers/render-automation';
import { renderInputsOfAudioNode } from '../helpers/render-inputs-of-audio-node';
export var createOscillatorNodeRendererFactory = function (createNativeOscillatorNode) {
    return function () {
        var nativeNode = null;
        var start = null;
        var stop = null;
        return {
            set start(value) {
                start = value;
            },
            set stop(value) {
                stop = value;
            },
            render: function (proxy, offlineAudioContext) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                var options;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (nativeNode !== null) {
                                return [2 /*return*/, nativeNode];
                            }
                            nativeNode = getNativeNode(proxy);
                            if (!!isOwnedByContext(nativeNode, offlineAudioContext)) return [3 /*break*/, 3];
                            options = {
                                channelCount: nativeNode.channelCount,
                                channelCountMode: nativeNode.channelCountMode,
                                channelInterpretation: nativeNode.channelInterpretation,
                                detune: nativeNode.detune.value,
                                frequency: nativeNode.frequency.value,
                                // @todo periodicWave is not exposed by the native node.
                                type: nativeNode.type
                            };
                            nativeNode = createNativeOscillatorNode(offlineAudioContext, options);
                            if (start !== null) {
                                nativeNode.start(start);
                            }
                            if (stop !== null) {
                                nativeNode.stop(stop);
                            }
                            return [4 /*yield*/, renderAutomation(proxy.context, offlineAudioContext, proxy.detune, nativeNode.detune)];
                        case 1:
                            _a.sent();
                            return [4 /*yield*/, renderAutomation(proxy.context, offlineAudioContext, proxy.frequency, nativeNode.frequency)];
                        case 2:
                            _a.sent();
                            return [3 /*break*/, 6];
                        case 3: return [4 /*yield*/, connectAudioParam(proxy.context, offlineAudioContext, proxy.detune)];
                        case 4:
                            _a.sent();
                            return [4 /*yield*/, connectAudioParam(proxy.context, offlineAudioContext, proxy.frequency)];
                        case 5:
                            _a.sent();
                            _a.label = 6;
                        case 6: return [4 /*yield*/, renderInputsOfAudioNode(proxy, offlineAudioContext, nativeNode)];
                        case 7:
                            _a.sent();
                            return [2 /*return*/, nativeNode];
                    }
                });
            }); }
        };
    };
};
