// standardized-audio-context/build/es2015/factories/analyser-node-renderer-factory.js
import * as tslib_1 from "tslib";
import { getNativeNode } from '../helpers/get-native-node';
import { isOwnedByContext } from '../helpers/is-owned-by-context';
import { renderInputsOfAudioNode } from '../helpers/render-inputs-of-audio-node';
export const createAnalyserNodeRendererFactory = (createNativeAnalyserNode) => {
    return () => {
        let nativeNode = null;
        return {
            render: (proxy, offlineAudioContext) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                if (nativeNode !== null) {
                    return nativeNode;
                }
                nativeNode = getNativeNode(proxy);
                // If the initially used nativeNode was not constructed on the same OfflineAudioContext it needs to be created again.
                if (!isOwnedByContext(nativeNode, offlineAudioContext)) {
                    const options = {
                        channelCount: nativeNode.channelCount,
                        channelCountMode: nativeNode.channelCountMode,
                        channelInterpretation: nativeNode.channelInterpretation,
                        fftSize: nativeNode.fftSize,
                        maxDecibels: nativeNode.maxDecibels,
                        minDecibels: nativeNode.minDecibels,
                        smoothingTimeConstant: nativeNode.smoothingTimeConstant
                    };
                    nativeNode = createNativeAnalyserNode(offlineAudioContext, options);
                }
                yield renderInputsOfAudioNode(proxy, offlineAudioContext, nativeNode);
                return nativeNode;
            })
        };
    };
};
