// standardized-audio-context/build/es2015/factories/audio-buffer-source-node-constructor.js
import { getNativeContext } from '../helpers/get-native-context';
const DEFAULT_OPTIONS = {
    buffer: null,
    channelCount: 2,
    channelCountMode: 'max',
    channelInterpretation: 'speakers',
    detune: 0,
    loop: false,
    loopEnd: 0,
    loopStart: 0,
    playbackRate: 1
};
export const createAudioBufferSourceNodeConstructor = (createAudioBufferSourceNodeRenderer, createAudioParam, createInvalidStateError, createNativeAudioBufferSourceNode, isNativeOfflineAudioContext, noneAudioDestinationNodeConstructor) => {
    return class AudioBufferSourceNode extends noneAudioDestinationNodeConstructor {
        constructor(context, options = DEFAULT_OPTIONS) {
            const nativeContext = getNativeContext(context);
            const mergedOptions = Object.assign({}, DEFAULT_OPTIONS, options);
            const nativeNode = createNativeAudioBufferSourceNode(nativeContext, mergedOptions);
            const isOffline = isNativeOfflineAudioContext(nativeContext);
            const audioBufferSourceNodeRenderer = (isOffline) ? createAudioBufferSourceNodeRenderer() : null;
            super(context, nativeNode, audioBufferSourceNodeRenderer);
            this._audioBufferSourceNodeRenderer = audioBufferSourceNodeRenderer;
            this._detune = createAudioParam(context, isOffline, nativeNode.detune);
            this._isBufferNullified = false;
            this._isBufferSet = false;
            this._nativeNode = nativeNode;
            // Bug #73: Edge, Firefox & Safari do not export the correct values for maxValue and minValue.
            this._playbackRate = createAudioParam(context, isOffline, nativeNode.playbackRate, 3.4028234663852886e38, -3.4028234663852886e38);
        }
        get buffer() {
            if (this._isBufferNullified) {
                return null;
            }
            return this._nativeNode.buffer;
        }
        set buffer(value) {
            // Bug #71: Edge does not allow to set the buffer to null.
            try {
                this._nativeNode.buffer = value;
            }
            catch (err) {
                if (value !== null || err.code !== 17) {
                    throw err;
                }
                // @todo Create a new internal nativeNode.
                this._isBufferNullified = (this._nativeNode.buffer !== null);
            }
            // Bug #72: Only Chrome, Edge & Opera do not allow to reassign the buffer yet.
            if (value !== null) {
                if (this._isBufferSet) {
                    throw createInvalidStateError();
                }
                this._isBufferSet = true;
            }
        }
        get onended() {
            return this._nativeNode.onended;
        }
        set onended(value) {
            this._nativeNode.onended = value;
        }
        get detune() {
            return this._detune;
        }
        get loop() {
            return this._nativeNode.loop;
        }
        set loop(value) {
            this._nativeNode.loop = value;
        }
        get loopEnd() {
            return this._nativeNode.loopEnd;
        }
        set loopEnd(value) {
            this._nativeNode.loopEnd = value;
        }
        get loopStart() {
            return this._nativeNode.loopStart;
        }
        set loopStart(value) {
            this._nativeNode.loopStart = value;
        }
        get playbackRate() {
            return this._playbackRate;
        }
        start(when = 0, offset = 0, duration) {
            this._nativeNode.start(when, offset, duration);
            if (this._audioBufferSourceNodeRenderer !== null) {
                this._audioBufferSourceNodeRenderer.start = (duration === undefined) ? [when, offset] : [when, offset, duration];
            }
        }
        stop(when = 0) {
            this._nativeNode.stop(when);
            if (this._audioBufferSourceNodeRenderer !== null) {
                this._audioBufferSourceNodeRenderer.stop = when;
            }
        }
    };
};
