// standardized-audio-context/build/es2015/factories/audio-node-constructor.js
import { EventTarget } from '../event-target';
import { AUDIO_NODE_STORE, AUDIO_PARAM_STORE, CONTEXT_STORE } from '../globals';
import { isAudioNode } from '../guards/audio-node';
import { cacheTestResult } from '../helpers/cache-test-result';
import { getAudioGraph } from '../helpers/get-audio-graph';
import { getAudioNodeConnections } from '../helpers/get-audio-node-connections';
import { getAudioParamConnections } from '../helpers/get-audio-param-connections';
import { getNativeContext } from '../helpers/get-native-context';
import { testAudioNodeDisconnectMethodSupport } from '../support-testers/audio-node-disconnect-method';
import { wrapAudioNodeDisconnectMethod } from '../wrappers/audio-node-disconnect-method';
const addAudioNode = (context, audioNode, audioNoderRender, nativeNode) => {
    const audioGraphOfContext = getAudioGraph(context);
    const inputs = [];
    for (let i = 0; i < nativeNode.numberOfInputs; i += 1) {
        inputs.push(new Set());
    }
    const audioNodeConnections = { inputs, outputs: new Set(), renderer: audioNoderRender };
    audioGraphOfContext.nodes.set(audioNode, audioNodeConnections);
    audioGraphOfContext.nodes.set(nativeNode, audioNodeConnections);
};
const addConnectionToAudioNode = (source, destination, output, input) => {
    const audioNodeConnectionsOfSource = getAudioNodeConnections(source);
    const audioNodeConnectionsOfDestination = getAudioNodeConnections(destination);
    audioNodeConnectionsOfSource.outputs.add([destination, output, input]);
    audioNodeConnectionsOfDestination.inputs[input].add([source, output]);
};
const addConnectionToAudioParam = (context, source, destination, output) => {
    const audioNodeConnections = getAudioNodeConnections(source);
    const audioParamConnections = getAudioParamConnections(context, destination);
    audioNodeConnections.outputs.add([destination, output]);
    audioParamConnections.inputs.add([source, output]);
};
const removeAnyConnection = (source) => {
    const audioNodeConnectionsOfSource = getAudioNodeConnections(source);
    for (const [destination] of Array.from(audioNodeConnectionsOfSource.outputs.values())) {
        if (isAudioNode(destination)) {
            const audioNodeConnectionsOfDestination = getAudioNodeConnections(destination);
            for (const connectionsToInput of audioNodeConnectionsOfDestination.inputs) {
                for (const connection of Array.from(connectionsToInput.values())) {
                    if (connection[0] === source) {
                        connectionsToInput.delete(connection);
                    }
                }
            }
        }
    }
    audioNodeConnectionsOfSource.outputs.clear();
};
const removeConnectionToAudioNode = (source, destination) => {
    const audioNodeConnectionsOfSource = getAudioNodeConnections(source);
    const audioNodeConnectionsOfDestination = getAudioNodeConnections(destination);
    for (const connection of Array.from(audioNodeConnectionsOfSource.outputs.values())) {
        if (connection[0] === destination) {
            audioNodeConnectionsOfSource.outputs.delete(connection);
        }
    }
    for (const connectionsToInput of audioNodeConnectionsOfDestination.inputs) {
        for (const connection of Array.from(connectionsToInput.values())) {
            if (connection[0] === source) {
                connectionsToInput.delete(connection);
            }
        }
    }
};
export const createAudioNodeConstructor = (createInvalidAccessError, isNativeOfflineAudioContext) => {
    return class AudioNode extends EventTarget {
        constructor(context, nativeNode, audioNodeRenderer) {
            super();
            this._context = context;
            this._nativeNode = nativeNode;
            const nativeContext = getNativeContext(context);
            // Bug #12: Firefox and Safari do not support to disconnect a specific destination.
            // @todo Make sure this is not used with an OfflineAudioContext.
            if (!isNativeOfflineAudioContext(nativeContext) && true !== cacheTestResult(testAudioNodeDisconnectMethodSupport, () => {
                return testAudioNodeDisconnectMethodSupport(nativeContext);
            })) {
                wrapAudioNodeDisconnectMethod(nativeNode);
            }
            AUDIO_NODE_STORE.set(this, nativeNode);
            addAudioNode(context, this, audioNodeRenderer, nativeNode);
        }
        get channelCount() {
            throw new Error('This needs to implemented.');
        }
        set channelCount(_) {
            throw new Error('This needs to implemented.');
        }
        get channelCountMode() {
            throw new Error('This needs to implemented.');
        }
        set channelCountMode(_) {
            throw new Error('This needs to implemented.');
        }
        get channelInterpretation() {
            return this._nativeNode.channelInterpretation;
        }
        set channelInterpretation(value) {
            this._nativeNode.channelInterpretation = value;
        }
        get context() {
            return this._context;
        }
        get numberOfInputs() {
            return this._nativeNode.numberOfInputs;
        }
        get numberOfOutputs() {
            return this._nativeNode.numberOfOutputs;
        }
        addEventListener(type, listener, // @todo EventListenerOrEventListenerObject | null = null,
        options) {
            return this._nativeNode.addEventListener(type, listener, options);
        }
        connect(destination, output = 0, input = 0) {
            const nativeContext = CONTEXT_STORE.get(this._context);
            if (isAudioNode(destination)) {
                // Bug #41: Only Chrome, Firefox and Opera throw the correct exception by now.
                if (this._context !== destination.context) {
                    throw createInvalidAccessError();
                }
                if (nativeContext === undefined) {
                    throw new Error('The native (Offline)AudioContext is missing.');
                }
                if (!isNativeOfflineAudioContext(nativeContext)) {
                    const nativeDestinationNode = AUDIO_NODE_STORE.get(destination);
                    if (nativeDestinationNode === undefined) {
                        throw new Error('The associated nativeNode is missing.');
                    }
                    if (nativeDestinationNode.inputs !== undefined) {
                        const inputs = nativeDestinationNode.inputs;
                        const nativeInputDestinationNode = inputs[input];
                        this._nativeNode.connect(nativeInputDestinationNode, output, input);
                    }
                    else {
                        this._nativeNode.connect(nativeDestinationNode, output, input);
                    }
                }
                addConnectionToAudioNode(this, destination, output, input);
                return destination;
            }
            if (nativeContext === undefined) {
                throw new Error('The native (Offline)AudioContext is missing.');
            }
            const nativeAudioParam = AUDIO_PARAM_STORE.get(destination);
            if (nativeAudioParam === undefined) {
                throw new Error('The associated nativeAudioParam is missing.');
            }
            try {
                this._nativeNode.connect(nativeAudioParam, output);
                // @todo Calling connect() is only needed to throw possible errors when the nativeContext is an OfflineAudioContext.
                if (isNativeOfflineAudioContext(nativeContext)) {
                    this._nativeNode.disconnect(nativeAudioParam, output);
                }
            }
            catch (err) {
                // Bug #58: Only Firefox does throw an InvalidStateError yet.
                if (err.code === 12) {
                    throw createInvalidAccessError();
                }
                throw err;
            }
            addConnectionToAudioParam(this._context, this, destination, output);
        }
        disconnect(destination) {
            const nativeContext = CONTEXT_STORE.get(this._context);
            if (nativeContext === undefined) {
                throw new Error('The native (Offline)AudioContext is missing.');
            }
            if (!isNativeOfflineAudioContext(nativeContext)) {
                if (destination === undefined) {
                    return this._nativeNode.disconnect();
                }
                const nativeDestinationNode = AUDIO_NODE_STORE.get(destination);
                if (nativeDestinationNode === undefined) {
                    throw new Error('The associated nativeNode is missing.');
                }
                if (nativeDestinationNode.inputs !== undefined) {
                    for (const input of nativeDestinationNode.inputs) {
                        this._nativeNode.disconnect(input);
                    }
                }
                else {
                    this._nativeNode.disconnect(nativeDestinationNode);
                }
            }
            if (destination === undefined) {
                removeAnyConnection(this);
            }
            else {
                removeConnectionToAudioNode(this, destination);
            }
        }
        removeEventListener(type, listener, // @todo EventListenerOrEventListenerObject | null = null,
        options) {
            return this._nativeNode.removeEventListener(type, listener, options);
        }
    };
};
