// standardized-audio-context/build/es2015/factories/media-element-audio-source-node-constructor.js
import { getNativeContext } from '../helpers/get-native-context';
const DEFAULT_OPTIONS = {
    channelCount: 2,
    channelCountMode: 'max',
    channelInterpretation: 'speakers'
};
export const createMediaElementAudioSourceNodeConstructor = (noneAudioDestinationNodeConstructor) => {
    return class MediaElementAudioSourceNode extends noneAudioDestinationNodeConstructor {
        constructor(context, options) {
            const nativeContext = getNativeContext(context);
            const mergedOptions = Object.assign({}, DEFAULT_OPTIONS, options);
            const nativeNode = nativeContext.createMediaElementSource(mergedOptions.mediaElement);
            super(context, nativeNode, null);
            // Bug #63: Edge & Firefox do not expose the mediaElement yet.
            this._mediaElement = mergedOptions.mediaElement;
            this._nativeNode = nativeNode;
        }
        get mediaElement() {
            return (this._nativeNode.mediaElement === undefined) ? this._mediaElement : this._nativeNode.mediaElement;
        }
    };
};
