// standardized-audio-context/build/es2015/factories/minimal-base-audio-context-constructor.js
import { EventTarget } from '../event-target';
import { CONTEXT_STORE } from '../globals';
export const createMinimalBaseAudioContextConstructor = (audioDestinationNodeConstructor) => {
    return class MinimalBaseAudioContext extends EventTarget {
        constructor(context, numberOfChannels) {
            super();
            CONTEXT_STORE.set(this, context);
            this._context = context;
            this._destination = new audioDestinationNodeConstructor(this, numberOfChannels);
        }
        get currentTime() {
            return this._context.currentTime;
        }
        get destination() {
            return this._destination;
        }
        get onstatechange() {
            return this._context.onstatechange;
        }
        set onstatechange(value) {
            this._context.onstatechange = value;
        }
        get sampleRate() {
            return this._context.sampleRate;
        }
        get state() {
            return this._context.state;
        }
    };
};
